src/xo-app/new/vm/index.js

```diff
@@ -1,214 +1,766 @@
+      } else {
+        cloudContent = state.customConfig
+      }
+    }
+
+    if (state.template.name_label === 'CoreOS') {
+      cloudContent = 'CoreOS'
+    }
+
+    const data = {
+      clone: state.fastClone,
+      existingDisks: state.existingDisks,
+      installation,
+      name_label: state.name_label,
+      template: state.template.id,
+      VDIs: state.VDIs,
+      VIFs: state.VIFs,
+      // TODO: To be added in xo-server
+      // vm.set parameters
+      CPUs: state.CPUs,
+      cpuWeight: state.cpuWeight,
+      name_description: state.name_description,
+      memory: state.memory,
+      pv_args: state.pv_args,
+      // Boolean: if true, boot the VM right after its creation
+      bootAfterCreate: state.bootAfterCreate,
+      /* String:
+       * - if 'CoreOS': vm.getCloudInitConfig --> vm.createCloudInitConfigDrive --> docker.register
+       * - else: vm.createCloudInitConfigDrive(..., cloudContent) --> vm.setBootOrder
+       */
+      cloudContent
+
+    }
+    createVm(data)
+  }
+
+  _selectPool = pool => {
+    this.setState({ pool })
+    this._reset()
+  }
+
+  _getIsInPool = createSelector(
+    () => this.state.pool,
+    pool => object => object.$pool === pool.id
+  )
+
+  _getSrPredicate = createSelector(
+    () => this.state.pool,
+    pool => disk => disk.$pool === pool.id && disk.content_type === 'user'
+  )
+
+  _initTemplate = template => {
+    if (!template) {
+      return this._reset()
+    }
+
+    const state = store.getState()
+
+    const existingDisks = {}
+    forEach(template.$VBDs, vbdId => {
+      const vbd = getObject(state, vbdId)
+      if (vbd.is_cd_drive) {
+        return
+      }
+      const vdi = getObject(state, vbd.VDI)
+      existingDisks[this.uniqueId] = {
+        name_label: vdi.name_label,
+        name_description: vdi.name_description,
+        size: vdi.size,
+        $SR: vdi.$SR
+      }
+    })
+
+    const VIFs = []
+    forEach(template.VIFs, vifId => {
+      const vif = getObject(state, vifId)
+      VIFs.push({
+        mac: vif.MAC,
+        network: vif.$network,
+        id: this.uniqueId
+      })
+    })
+
+    this.setState({
+      // infos
+      name_label: template.name_label,
+      template,
+      name_description: template.name_description || this.state.name_description || '',
+      // performances
+      memory: template.memory.size,
+      CPUs: template.CPUs.number,
+      cpuWeight: 1,
+      // installation
+      installMethod: template.install_methods && template.install_methods[0] || this.state.installMethod,
+      // interfaces
+      VIFs,
+      // disks
+      existingDisks,
+      VDIs: map(template.template_info.disks, disk => ({ ...disk, device: this.uniqueId }))
+    }, () => forEach(this.state, (element, key) => {
+      !isArray(element) && this._setRef(key, element)
+    }))
+  }
+
+  _addVdi = () => this.setState({ VDIs: concat(this.state.VDIs, {
```

Comment:
Use ES6 spread instead of concat: `VDIs: [ ...this.state.VDIs, { device: String(this.uniqueId), type: 'system' } ]`.